0027-interledger-protocol-4/0027-interledger-protocol-4.md

```diff
@@ -94,7 +94,7 @@ In the case of post-funded accounts, senders transfer value to their connector a
 
 ### Packet Format
 
-ILPv4 packets SHOULD be encoded using the ASN.1 Octet Encoding Rules.
+ILPv4 packets MUST be encoded using the ASN.1 Canonical Octet Encoding Rules.
```

Comment:
Are we sure we want this to read `MUST`? 

For example, if two peers (`A` & `B`) wanted to exchange ILP packet information using some non-OER encoding (e.g., HTTP headers), and then `B` and `C` wanted to exchange packets using OER encoded "packets", would this not still be considered "ILP"?

Reply:
Speaking for myself, no, not sure :) In that case, seems neither SHOULD or MUST are really appropriate? 

Just reading this:
https://medium.com/interledger-blog/thoughts-on-scaling-interledger-connectors-7e3cad0dab7f

It's interesting how much information in ILP is implied by the connection

Reply:
I think SHOULD is the right way to say it.

> 3. SHOULD   This word, or the adjective "RECOMMENDED", mean that there
   may exist valid reasons in particular circumstances to ignore a
   particular item, but the full implications must be understood and
   carefully weighed before choosing a different course.

From [RFC 2119](https://tools.ietf.org/html/rfc2119)

Reply:
Seems so

Reply:
The [robustness principle](https://en.wikipedia.org/wiki/Robustness_principle) says a compliant implementation MUST produce correct encodings but MAY accept other, invalid encodings.

I think it should say that you MUST encode with Canonical OER but you MAY accept packets that are not encoded canonically.

Reply:
That's a good, and in my view, different point. 

It seemed like there were a lot of good reasons enumerated for using C-OER. If you have a well defined standard for canonical encodings it makes sense to mandate them to me. 

Otherwise, in practice you'll have to define (read: arise ad-hoc) what are semantically equivalent/meaningful encodings. ([See criticisms](https://en.wikipedia.org/wiki/Robustness_principle#Criticism)) I'm not advocating to "fail fast" and crash if you are being sent non-conformant packets by malevolent actors though. 

I couldn't quite articulate what I meant before when I said "seems neither are really appropriate". You aren't really saying much with SHOULD other than "this is the default encoding format, but we might support others", though in a manner contorting to conform to FORMAL WORDING. 

wrt: encoding packets in HTTP headers, I'm not really up-to-speed with why you would want to do that, but sticking to one way of doing things seems wise. 

I'll appeal to Tim Peters
```
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
```

Do we know any Dutch ?

Reply:
From my perspective, I'm good with [either](https://github.com/interledger/rfcs/pull/523#discussion_r268683409) / [option](https://github.com/interledger/rfcs/pull/523#discussion_r274223268) proposed by @emschwartz  or @justmoon - I think both of them leave room for alternate implementations, which was my main concern.